presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergMetadata.java

```diff
@@ -662,4 +702,335 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
+                        .withFormat(table.getFileFormat())
+                        .withMetrics(task.getMetrics().metrics());
+
+                if (!icebergTable.spec().fields().isEmpty()) {
+                    String partitionDataJson = task.getPartitionDataJson()
+                            .orElseThrow(() -> new VerifyException("No partition data for partitioned table"));
+                    builder.withPartition(PartitionData.fromJson(partitionDataJson, partitionColumnTypes));
+                }
+
+                appendFiles.appendFile(builder.build());
+            }
+
+            // Update the map that tracks tables on which the materialized view depends and the corresponding snapshot ids of the table
+            String tableSnapshotIDString = Joiner.on(",").withKeyValueSeparator("=").join(transactionHandle.getTableSnapshotIdMap());
+            appendFiles.set(DEPENDS_ON_TABLES, tableSnapshotIDString);
+            appendFiles.commit();
+        }
+        transaction.commitTransaction();
+        return Optional.of(new HiveWrittenPartitions(commitTasks.stream()
+            .map(CommitTaskData::getPath)
+            .collect(toImmutableList())));
+    }
+
+    private boolean isAMaterializedView(Table table)
+    {
+        if (table.getTableType().equals(org.apache.hadoop.hive.metastore.TableType.VIRTUAL_VIEW.name()) &&
+                "true".equals(table.getParameters().get(PRESTO_VIEW_FLAG)) &&
+                table.getParameters().containsKey(STORAGE_TABLE)) {
+            return true;
+        }
+        return false;
+    }
+
+    private boolean isAMaterializedView(ConnectorSession session, SchemaTableName schemaTableName)
+    {
+        final HiveIdentity identity = new HiveIdentity(session);
+        if (metastore.getTable(identity, schemaTableName.getSchemaName(), schemaTableName.getTableName()).isPresent()) {
+            Table table = metastore.getTable(identity, schemaTableName.getSchemaName(), schemaTableName.getTableName()).get();
+            return isAMaterializedView(table);
+        }
+        return false;
+    }
+
+    public static List<ConnectorMaterializedViewDefinition.Column> icebergSchemaToMaterializedViewCols(TypeManager typeManager,
+            List<io.prestosql.plugin.hive.metastore.Column> schema)
+    {
+        List<ConnectorMaterializedViewDefinition.Column> columns = new ArrayList<>();
+        for (io.prestosql.plugin.hive.metastore.Column field : schema) {
+            ConnectorMaterializedViewDefinition.Column column =
+                    new ConnectorMaterializedViewDefinition.Column(field.getName(), field.getType().getType(typeManager).getTypeId());
+            columns.add(column);
+        }
+        return columns;
+    }
+
+    @Override
+    public Optional<ConnectorMaterializedViewDefinition> getMaterializedView(ConnectorSession session, SchemaTableName viewName)
+    {
+        if (!isAMaterializedView(session, viewName)) {
```

Comment:
This is inherently racey since the definition of the table might change between the time that we check it and when we use it later. We should fetch the table metadata, then check if it's a materialized view.

Reply:
ok, I see the race condition .. code checks if an object is a materialized view .. before it gets the metadata of the materialized view somebody goes and drops it and creates a table with the same name .. access to metadata is stale and incorrect.
Fixed the order.